### PR TITLE
PUBDEV-3416: Implements assignment to a sliced frame

### PR DESCRIPTION
**Pull request by michalkurka**
needs more tests

---
This change is [Reviewable](https://reviewable.io/reviews/h2oai/h2o-3/261)

### CODE REVIEW

**Comment by st-pasha** [comment]
Looks good, but see [PUBDEV-3487] and [PUBDEV-3488]

File: h2o-core/src/main/java/water/rapids/ast/params/AstNumList.java [hunk 1]
```diff
@@ -263,18 +263,34 @@ public boolean isEmpty() {
...
-    int idx = Arrays.binarySearch(_bases, v);
+    int idx = findBase(v);
     if (idx >= 0) return true;
     idx = -idx - 2;  // See Arrays.binarySearch; returns (-idx-1), we want +idx-1  ... if idx == -1 => then this transformation has no effect
     if (idx < 0) return false;
     assert _bases[idx] < v;     // Sanity check binary search, AND idx >= 0
     return v < _bases[idx] + _cnts[idx] * _strides[idx];
   }
 
+  public long index(long v) {
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/rapids/ast/params/AstNumList.java]: Method comment would be great!

File: h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java [hunk 2]
```diff
@@ -128,11 +128,39 @@ private void assign_frame_frame(Frame dst, int[] cols, AstNumList rows, Frame sr
       return;
     }
     // Handle large case
-    throw H2O.unimpl();
+    Vec[] vecs = ses.copyOnWrite(dst, cols);
+    Vec[] vecs2 = new Vec[cols.length]; // Just the selected columns get updated
+    for (int i = 0; i < cols.length; i++)
+      vecs2[i] = vecs[cols[i]];
+    rows.sort();                // Side-effect internal sort; needed for fast row lookup
+    new RowSliceTask(rows) {
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java]: I would be careful here with the anonymous class - it captures the surrounding context , so you are probably distributing the `vecs`, `vecs2`, `rows`.

File: h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java [hunk 3]
```diff
@@ -162,25 +190,14 @@ private void assign_frame_scalar(Frame dst, int[] cols, final AstNumList rows, f
     for (int i = 0; i < cols.length; i++)
       vecs2[i] = vecs[cols[i]];
     rows.sort();                // Side-effect internal sort; needed for fast row lookup
-    new MRTask() {
+    new RowSliceTask(rows) {
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java]: The same as above - it is probably better to have a method returning the instance of task, or have a static type.

File: h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java [hunk 4]
```diff
@@ -245,4 +251,32 @@ public void map(Chunk[] cs, NewChunk[] ncs) {
     }.doAll(cols.length, Vec.T_NUM, new Frame(dst).add(rows)).outputFrame();
     assign_frame_frame(dst, cols, new AstNumList(0, dst.numRows()), src2, ses);
   }
+
+  private static abstract class RowSliceTask extends MRTask {
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java]: Minor typing: Should be `extends MRTask<RowSliceTask>`

File: h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java (hunk 2 above)
Comment (michalkurka) [inline h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java]: These are small objects, it should not matter much. The same pattern was used by Cliff, so I wasn't too worried about copying it.

Do you think I should extract them into static classes?

File: h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java (hunk 4 above)
Comment (michalkurka) [inline h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java]: Will fix, thank you.

File: h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java (hunk 2 above)
Comment (mmalohlava) [inline h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java]: It is dangerous pattern for future code updates. If somebody modifies the context and put there some expensive object and does not check the rest of code.

There are two solutions;
  1) separated static type
  2) a builder method: 
      ```java
        static public RowSliceTask createXXX(int[] rows) { return new RowSliceTask(rows) {
          @Override void mapChunkSlice() { ... }
        }
       ```

Comment (michalkurka) [inline h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java]: Fixed.

File: h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java (hunk 3 above)
Comment (michalkurka) [inline h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java]: Fixed.

File: h2o-core/src/main/java/water/rapids/ast/params/AstNumList.java (hunk 1 above)
Comment (michalkurka) [inline h2o-core/src/main/java/water/rapids/ast/params/AstNumList.java]: Fixed.

**Pull request merged**